File: #1140.stone-game-ii.cpp
class Solution {
public:
    int stoneGameII(vector<int>& piles) {
        
        memset(dp, -1, sizeof(dp));
        return helper(0, true, 1, piles);
    }

private:
    int dp[101][203][2];

    int helper(int node, int player,
        int m, vector<int>&piles) {

        int size = piles.size();
        if (node >= size) {
            return 0;
        }
        
        if (dp[node][m][player] != -1) {
            return dp[node][m][player];
        }

        if (player) {
            int take = 0;
            int result = INT_MIN;
            for (int i = node ; i < node + 2 * m and i < size; i++ ) {
                take += piles[i];
                result = std::max(result,
                    take + helper(i+1, !player, std::max(m, i-node + 1), piles)); 
            }
            return dp[node][m][player] = result;
        }

        int take = 0;
        int result = INT_MAX;
        for (int i = node ; i < node + 2 * m and i < size; i++) {
            result = std::min(result,
                helper(i + 1, !player, std::max(m, i - node + 1), piles));
        }
        return dp[node][m][player] = result;
    }
};

File: #1140.stone-game-ii.js
/**
 * @param {number[]} piles
 * @return {number}
 */
var stoneGameII = function(piles) {
    let n = piles.length;
    let sum = new Array(n).fill(0);
    for (let i = n - 1; i >= 0; i--) {
        sum[i] = i == n - 1 ? piles[i] : piles[i] + sum[i + 1]; 
    }

    let arr = new Array(n + 1)
        .fill(0)
        .map(el => new Array(n + 1).fill(0));
    for (let i = n - 1; i >= 0; i--) {
        for (let j = 1; j <= n; j++) {
            for (let x = 1; x <= 2 * j && i + x <= n; x++) {
                arr[i][j] = Math.max(arr[i][j],
                    sum[i] - arr[i + x][Math.max(j, x)]);
            }
        }
    }
    return arr[0][1]; 
};
